book/src/leader-validator-transition.md

```diff
@@ -0,0 +1,157 @@
+        end_ticks: u64,
+        sender: Sender<Vec<Entry>>,
+	    bank_state: Arc<BankState>,
+        exit_signal: Sender<()>);
+
+    /// Record transactions /// Error is returned if end_ticks is reached first
+    pub fn record(&self, mixin: Hash, txs: Vec<Transaction>) -> Result<()>;
+} 
+```
+
+
+## Fullnode Loop
+
+This object manages the transition between modes.  The main idea is that once a
+ledger is replayed, the validator can run until there is proof that it can be a
+leader.  The Leader can then run and record its transactions.
+
+``` 
+/// * forks - the Forks generated by replaying the ledger
+/// * leader_scheduler - the leader_scheduler generated by replaying the ledger 
+/// * ncp -  the network
+
+let main(forks: Forks, leader_scheduler: LeaderScheduler, ncp: Ncp) {
+    let (exit_receiver, exit_signal) = channel(); 
+    loop {
+        // run the validator loop first
+        let my_slot = leader_scheduler.get_scheduled_tick_height(my_id);
+        let last_fork = forks.latest_fork();
+    
+        // the generator will exit immediatly if tick_height == my_slot.start
+        assert!(last_fork.tick_height <= my_slot.start,
+            "start is out of sync, replay the ledger!");
+        let generator = PohGenerator::new(last_fork.last_id, last_fork.tick_height,
+            my_slot.start, exit_signal);
+    
+        // operate the validator, this runs until my_slot.start is reached by the
+        // generator
+        let validator = Validator::new(forks, leader_scheduler, generator, ncp);
+    
+        // wait for validator to finish
+        exit_receiver.recv();
+        // the bank and forked state is outside thie loop;
+        validator.exit().
+    
+        // these entries connect to my_slot
+        let poh_entries = generator.entries();
+    
+        // make sure this is true
+        assert!(poh_entries.last().unwrap().tick_height == my_slot.start,
+            "generator didn't reach my scheduled height, abort!");
+    
+        // get the slot these entries connect to
+        let starting_slot = poh_entries
+                            .first()
+                            .map(|e| e.slot_index()  - 1)
+                            .unwrap_or(my_slot.slot_id - 1);
+    
+        // create a fork from the start to my slot
+        let bank_state = forks.init(my_slot.slot_id, starting_slot);
+    
+        // operate as leader
+        // recorder will register ticks into the bank_state
+        let recorder = PohRecorder::new(poh_entries, my_slot.end, bank_state,
+            exit_signal);
+        let leader = Leader::new(bank_state, recorder, ncp);
+        exit_receiver.recv();
+        leader.exit();
+    
+        // Finalize the bank_state, send the vote as a validator.
+        bank_state.finalize();
+    }
+}
+```
+
+## BankState
+
+BankState operates over a specific slot.  Once the final tick has been
+registered the state becomes finalized and further writes will error out.
```

Comment:
This looks like a confusing use of the term *finalized*. Under current terminology, it doesn't look to be finalized or even confirmed.